File: AIM/AIM/UI/Views/TrackListViewController.m
//
//  TrackListViewController.m
//  AIM
//
//  Created by Oskar Szydlowski on 05.03.2018.
//  Copyright © 2018 Oskar Szydlowski. All rights reserved.
//

#import "TrackListViewController.h"

#import "TrackCell.h"

@implementation TrackListViewController

static NSString *cellIdentifier = @"TrackCell";
static CGFloat rowHeight = 90.0;


- (instancetype)initWithPresenter:(TrackListPresenter *)presenter {
    self = [super init];

    if (self != nil) {
        self.presenter = presenter;
        return self;
    }
    return nil;
}

- (void)loadView {
    [super loadView];
    self.view = [[UIView new]initWithFrame:[[UIScreen mainScreen] bounds]];
    self.view.backgroundColor = [UIColor redColor];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupTableView];
    [self fetchOnAirInfo];
}

- (void)setupTableView {
    self.tableView = [UITableView new];
    self.tableView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.tableView];
    [self.view addConstraints:@[
                           [self.tableView.topAnchor constraintEqualToAnchor:self.view.topAnchor],
                           [self.tableView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
                           [self.tableView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
                           [self.tableView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor]
    ]];
    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    [self.tableView registerClass:[TrackCell self] forCellReuseIdentifier:cellIdentifier];
}

// MARK: - UITableViewDelegate

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return rowHeight;
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    if ([self.presenter shouldDisplayHeader]) {
        return rowHeight;
    } else {
        return 0.0;
    }
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    if ([self.presenter shouldDisplayHeader]) {
        StationInfoView *headerView = [StationInfoView new];
        [self.presenter configureHeaderView:headerView];
        return headerView;
    }
    return nil;
}

//MARK: - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.presenter numberOfTracks];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    TrackCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    [self.presenter configureCell:cell atIndex:indexPath.row];
    return cell;
}

// MARK: - TrackListView

- (void) refreshView {
    [self.tableView reloadData];
}

- (void) fetchOnAirInfo {
    [self.presenter fetchOnAirInfo];
}

@end

File: AIM/AIM/UI/Views/TrackListViewController.swift
//
//  TrackListViewController.swift
//  AIM
//
//  Created by Oskar Szydlowski on 28.02.2018.
//  Copyright © 2018 Oskar Szydlowski. All rights reserved.
//

import UIKit

class TrackListViewController: UIViewController, UITableViewDelegate, UITableViewDataSource, TrackListView {
    
    private let presenter: TrackListPresenter
    
    private let tableView = UITableView()
    
    private let rowHeight: CGFloat = 90.0
    
    init(presenter: TrackListPresenter) {
        self.presenter = presenter
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupTableView()
        fetchOnAirInfo()
    }
    
    private func setupTableView() {
        view.addSubview(tableView, with: [
            tableView.topAnchor.constraint(equalTo: view.topAnchor),
            tableView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            tableView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
        ])
        
        tableView.dataSource = self
        tableView.delegate = self
        tableView.register(TrackCell.self)        
    }
    
    // MARK: - UITableViewDelegate
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return rowHeight
    }
    
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return presenter.shouldDisplayHeader() ? rowHeight : 0.0
    }
    
    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        if presenter.shouldDisplayHeader() {
            let headerView = StationInfoView()
            presenter.configureHeaderView(headerView)
            return headerView
        }

        return nil
    }
    
    //MARK: - UITableViewDataSource
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return presenter.numberOfTracks()
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell: TrackCell = tableView.dequeueReusableCell(for: indexPath)
        presenter.configureCell(cell, at: indexPath.row)
        return cell
    }
    
    // MARK: - TrackListView
    
    func refreshView() {
        tableView.reloadData()
    }
    
    @objc private func fetchOnAirInfo() {
        presenter.fetchOnAirInfo()
    }
}
